datastream-server-restli/src/test/java/com/linkedin/datastream/server/TestCoordinator.java

```diff
@@ -1348,6 +1348,77 @@ public void testValidateDatastreamsUpdate() throws Exception {
 
+  @Test
+  public void testDatastreamCreateUpdateDelete() throws Exception {
+    String testCluster = "testDatastreamCreateUpdateDelete";
+    String connectorType = "connectorType";
+
+    TestHookConnector connector1 = new TestHookConnector("connector1", connectorType);
+    TestHookConnector connector2 = new TestHookConnector("connector2", connectorType);
+
+    Coordinator coordinator1 = createCoordinator(_zkConnectionString, testCluster);
+    coordinator1.addConnector(connectorType, connector1, new BroadcastStrategy(Optional.empty()), false,
+        new SourceBasedDeduper(), null);
+    coordinator1.start();
+
+    Coordinator coordinator2 = createCoordinator(_zkConnectionString, testCluster);
+    coordinator2.addConnector(connectorType, connector2, new BroadcastStrategy(Optional.empty()), false,
+        new SourceBasedDeduper(), null);
+    coordinator2.start();
+
+    ZkClient zkClient = new ZkClient(_zkConnectionString);
+
+    Datastream[] list =
+        DatastreamTestUtils.createAndStoreDatastreams(zkClient, testCluster, connectorType, "datastream1");
+    Datastream datastream = list[0];
+    LOG.info("Created datastream: {}", datastream);
+
+    // wait for datastream to be READY
+    PollUtils.poll(() -> DatastreamTestUtils.getDatastream(zkClient, testCluster, "datastream1")
+        .getStatus()
+        .equals(DatastreamStatus.READY), 1000, WAIT_TIMEOUT_MS);
+    datastream = DatastreamTestUtils.getDatastream(zkClient, testCluster, datastream.getName());
+    assertConnectorAssignment(connector1, WAIT_TIMEOUT_MS, datastream.getName());
+
+    // ZK chooses instance with lower index count as a leader from the cluster
+    // coordinator1 has to be leader as it gets created first with lower index counter
+    Assert.assertTrue(coordinator1.getIsLeader().getAsBoolean());
+
+    // connector1 is assigned on coordinator1 which is a leader
+    // and post datastream create/delete or state chance method should be invoked only for a leader instance
+    Assert.assertEquals(connector1.getPostDSStatechangeActionInvokeCount(), 1);
+
+    // update datastream
+    datastream.getMetadata().put("key", "value");
+    datastream.getSource().setConnectionString("newSource");
+
+    LOG.info("Updating datastream: {}", datastream);
+    CachedDatastreamReader datastreamCache = new CachedDatastreamReader(zkClient, testCluster);
+    ZookeeperBackedDatastreamStore dsStore = new ZookeeperBackedDatastreamStore(datastreamCache, zkClient, testCluster);
+    DatastreamResources datastreamResources = new DatastreamResources(dsStore, coordinator1);
+    datastreamResources.update(datastream.getName(), datastream);
+
+    assertConnectorReceiveDatastreamUpdate(connector1, datastream);
+    Assert.assertTrue(PollUtils.poll(() -> connector1.getPostDSStatechangeActionInvokeCount() == 2, 200,
+        WAIT_TIMEOUT_MS));
+    Assert.assertTrue(PollUtils.poll(() -> connector2.getPostDSStatechangeActionInvokeCount() == 1, 200,
+        WAIT_TIMEOUT_MS));
```

Comment:
Why should connector2 also publish an event? If there are 100 hosts and the tasks are assigned to all of them, an update operation will result in 100 events.